File: e2e-testing/page_objects/DetailPage.js
/**
 * @file DetailPage.js
 * @description Page Object Model for Stock Opname detail modal and history logs
 * @author Hafidz (QA Engineer 4)
 */

const BasePage = require("./BasePage");

class DetailPage extends BasePage {
  constructor(page) {
    super(page);
    this.successToast = 'div:has-text("Sesi opname ditutup")';
    this.historyTable = ".w-full.flex.flex-col table";
    this.detailModalTitle = 'h3:has-text("Detail Opname")';
    this.viewDetailBtn = 'button[title="Lihat Detail"]';
  }

  // Menunggu notifikasi sukses muncul
  async verifySuccessNotification() {
    // Menunggu toast react-hot-toast/notify muncul
    const toast = this.page.locator('div[role="status"], div:has-text("Sesi opname")').first();
    await toast.waitFor({ state: "visible", timeout: 10000 });
    return await toast.isVisible();
  }

  // Membuka modal detail sesi opname terbaru (paling atas di tabel riwayat)
  async openLatestOpnameDetail() {
    // Cari tombol aksi "Lihat Detail" pertama di tabel riwayat
    const firstViewBtn = this.page.locator(this.viewDetailBtn).first();
    await firstViewBtn.waitFor({ state: "visible", timeout: 15000 });
    await firstViewBtn.click();
  }

  // Memverifikasi isi data item di dalam modal detail
  async verifyDetailItem(itemName, expectedQuantity, expectedNotes) {
    // Pastikan modal detail terbuka
    await this.page.waitForSelector(this.detailModalTitle);
    
    // Temukan baris tabel di modal untuk item tertentu
    const row = this.page.locator("tr").filter({ has: this.page.locator("p", { hasText: itemName }) });
    await row.waitFor({ state: "visible" });
    
    // Ambil kolom kuantitas aktual (kolom ke-3) dan catatan (kolom ke-5)
    // Berdasarkan markup, urutan kolom: Nama Item (1), Sistem (2), Aktual (3), Selisih (4), Catatan (5)
    const physicalCell = row.locator("td").nth(2);
    const notesCell = row.locator("td").nth(4);
    
    const actualQuantity = await physicalCell.textContent();
    const actualNotes = await notesCell.textContent();
    
    return {
      quantityMatches: actualQuantity.trim() === expectedQuantity.toString(),
      notesMatches: actualNotes.trim().includes(expectedNotes) || actualNotes.trim() === expectedNotes
    };
  }

  // Menutup modal detail
  async closeDetailModal() {
    const closeBtn = this.page.locator('button:has-text("Batal"), button[aria-label="Close"], svg').first();
    await closeBtn.click();
  }
}

module.exports = DetailPage;

File: e2e-testing/src/test/java/com/autoservice/pages/DetailPage.java
package com.autoservice.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.List;

/**
 * @file DetailPage.java
 * @description Page Object Model for Stock Opname detail modal and history logs in Java
 * @author Hafidz (QA Engineer 4)
 */
public class DetailPage extends BasePage {
    private final By successToast = By.xpath("//*[contains(text(), 'Sesi opname ditutup')]");
    private final By viewDetailBtn = By.cssSelector("button[title='Lihat Detail']");
    private final By detailModalTitle = By.xpath("//h3[contains(., 'Detail Opname')]");

    public DetailPage(WebDriver driver) {
        super(driver);
    }

    public boolean verifySuccessNotification() {
        return isElementVisible(successToast);
    }

    public void openLatestOpnameDetail() {
        wait(1000);
        waitForElementVisible(viewDetailBtn);
        List<WebElement> buttons = driver.findElements(viewDetailBtn);
        if (!buttons.isEmpty()) {
            buttons.getFirst().click();
        }
    }

    public boolean verifyDetailItem(String itemName, int expectedQuantity, String expectedNotes) {
        waitForElementVisible(detailModalTitle);
        
        By rowLocator = By.xpath("//tr[contains(., '" + itemName + "')]");
        WebElement row = driver.findElement(rowLocator);
        
        WebElement physicalCell = row.findElement(By.xpath("./td[3]"));
        WebElement notesCell = row.findElement(By.xpath("./td[5]"));
        
        String actualQuantity = physicalCell.getText().trim();
        String actualNotes = notesCell.getText().trim();
        
        boolean quantityMatches = actualQuantity.equals(String.valueOf(expectedQuantity));
        boolean notesMatches = actualNotes.contains(expectedNotes) || actualNotes.equals(expectedNotes);
        
        return quantityMatches && notesMatches;
    }

    public void closeDetailModal() {
        By closeBtnLocator = By.xpath("//button[contains(., 'Batal') or contains(., 'Tutup') or contains(@aria-label, 'Close') or contains(., '×')]");
        if (isElementVisible(closeBtnLocator)) {
            click(closeBtnLocator);
        } else {
            wait(500);
        }
    }
}
